### ISSUE TITLE
Memory usage of qdrant on the desktop

### ISSUE THREAD

**Issue by amircodota** [qdrant/qdrant#5184]
Hi folks

We are using qdrant for a desktop ai assistant.

We are running qdrant on the desktop, indexing many documents and allowing to find relevant snippets.
Overall we are very happy with qdrant - both quality and performance wise - running it on windows, mac and linux.

We did run into an issue of memory usage.
When the user indexes a big body of documents, qdrant requires a lot of disk and memory - that's to be expected. 
However, it seems that even after qdrant is stopped and started again, it immediately loads all of its collections into memory. Even the ones we do not use or need at the moment.

We are running with on_disk: true, and I'm aware that qdrant uses mmap, so if it runs into memory pressure some of this memory will go down to disk.
However this causes slow downs for our users - it causes excessive paging to happen, and sometimes slows the desktop machine down considerably.

It would be great if we can somehow alter qdrant's memory behaviour. My knowledge of mmaps is somewhat limited, but to my understanding, it only reads pages into memory when needed. So I suspect qdrant reads the entire contents of the collection on startup, loading everything into memory.

I'm aware this is not qdrant's main use case, but it would be great if there was some way to:
1. Make qdrant read pages only when needed
2. Alternatively, limit qdrant's memory usage so it does not pass a specified threshold (there might be ways to achieve that using OS level limits, but those are highly platform dependant and complex)

**Comment by timvisee** [comment]
> it immediately loads all of its collections into memory. Even the ones we do not use or need at the moment.

Correct. We have no plans at this time to add dynamic loading for collections. Meaning we load/unload automatically as needed.

We try to warm up all memory mapped pages so that first requests will be quick. What's why you might be seeing expensive IO.

In [Qdrant 1.11.4](https://github.com/qdrant/qdrant/releases/tag/v1.11.4) we made this process more efficient by asking the kernel to do this for us. However, this does not work on Windows and it's still using the naive approach there.

> 1. Make qdrant read pages only when needed

First requests using this approach are extremely slow, and likely timeout. That's why we don't provide a toggle for this at this time.

Qdrant uses all memory available to it, as that's preferred when deploying it on a server. It helps make Qdrant as fast as possible, because unused memory is wasted memory. What I can suggest now is to limit the memory available to Qdrant, basically what you mention in point 2. We recommend to use tools available on the platform to achieve this. For example, set a memory limit on the Docker container.

Could you try the above and see if that improves the situation? I understand this answer may not be very satisfactory, but ideally we'd use an approach which is already supported.
